src/transforms/aggregate.js

```diff
@@ -0,0 +1,283 @@
+    var arrayAttrs = {};
+    for(i = 0; i < arrayAttrArray.length; i++) arrayAttrs[arrayAttrArray[i]] = 1;
+
+    var groups = coerce('groups');
+
+    if(!Array.isArray(groups)) {
+        if(!arrayAttrs[groups]) {
+            transformOut.enabled = false;
+            return;
+        }
+        arrayAttrs[groups] = 0;
+    }
+
+    var aggregationsIn = transformIn.aggregations;
+    var aggregationsOut = transformOut.aggregations = [];
+
+    if(aggregationsIn) {
+        for(i = 0; i < aggregationsIn.length; i++) {
+            var aggregationOut = {};
+            var array = Lib.coerce(aggregationsIn[i], aggregationOut, attrs.aggregations, 'array');
+            var func = Lib.coerce(aggregationsIn[i], aggregationOut, attrs.aggregations, 'func');
+
+            // add this aggregation to the output only if it's the first instance
+            // of a valid array attribute - or an unused array attribute with "count"
+            if(array && (arrayAttrs[array] || (func === 'count' && arrayAttrs[array] === undefined))) {
+                arrayAttrs[array] = 0;
+                aggregationsOut.push(aggregationOut);
+            }
+        }
+    }
+
+    // any array attributes we haven't yet covered, fill them with the default aggregation
+    for(i = 0; i < arrayAttrArray.length; i++) {
+        if(arrayAttrs[arrayAttrArray[i]]) {
+            aggregationsOut.push({
+                array: arrayAttrArray[i],
+                func: attrs.aggregations.func.dflt
+            });
+        }
+    }
+
+    return transformOut;
+};
+
+
+exports.calcTransform = function(gd, trace, opts) {
+    if(!opts.enabled) return;
+
+    var groups = opts.groups;
+
+    var groupArray = Lib.getTargetArray(trace, {target: groups});
+    if(!groupArray) return;
+
+    var i, vi, groupIndex;
+
+    var groupIndices = {};
+    var groupings = [];
+    for(i = 0; i < groupArray.length; i++) {
+        vi = groupArray[i];
+        groupIndex = groupIndices[vi];
+        if(groupIndex === undefined) {
+            groupIndices[vi] = groupings.length;
+            groupings.push([i]);
+        }
+        else groupings[groupIndex].push(i);
+    }
+
+    var aggregations = opts.aggregations;
+
+    for(i = 0; i < aggregations.length; i++) {
+        aggregateOneArray(gd, trace, groupings, aggregations[i]);
+    }
+
+    if(typeof groups === 'string') {
+        aggregateOneArray(gd, trace, groupings, {array: groups, func: 'first'});
+    }
+};
+
+function aggregateOneArray(gd, trace, groupings, aggregation) {
+    var attr = aggregation.array;
+    var targetNP = Lib.nestedProperty(trace, attr);
+    var arrayIn = targetNP.get();
+    var conversions = Axes.getDataConversions(gd, trace, attr, arrayIn);
+    var func = getAggregateFunction(aggregation.func, conversions);
+
+    var arrayOut = new Array(groupings.length);
+    for(var i = 0; i < groupings.length; i++) {
+        arrayOut[i] = func(arrayIn, groupings[i]);
```

Comment:
Good point - I think for the moment I'll leave it as is, all the existing aggregations would be fairly easy to replicate online (and yes, `'last'` is easy - perhaps even easier than `'first'`, you just always replace the output with the new value) but some of the others we've listed in the comments might be trickier (median and mode, in particular). I'll keep this in mind though as a potential performance gain for later.

To note from a private convo with @rreusser - the case where this is most important is when you have a large number of small groups - thousands of groups aggregating just a couple of items each - in which case creating all the little `groupings` arrays entails significant overhead. If we see unreasonable drag in this case, switching to online algorithms would be the fix.